src/cart/crt_rpc.h

```diff
@@ -130,8 +130,10 @@ struct crt_rpc_priv {
 	crt_rpc_t		crp_pub; /* public part */
 	/* link to crt_ep_inflight::epi_req_q/::epi_req_waitq */
 	d_list_t		crp_epi_link;
-	/* tmp_link used in crt_context_req_untrack */
-	d_list_t		crp_tmp_link;
+	/* link for temp list used during timeout processing */
+	d_list_t                 crp_tmp_link_timeout;
+	/* link for temp list used for wait q processing */
+	d_list_t                 crp_tmp_link_submit;
```

Comment:
fundamentally I'm not entirely sure why this requires two separate lists (even if this solves the issue here :) ). Also why do RPCs that have timed out or got canceled need to be on a list at all ?

Reply:
those lists in question are very temporary and are just used to be able to process rpcs outside of big locks.

On a sender on each crt_progress we go through all tracked rpcs and see which ones should be timed out by now. Those we add to the timeout list and then process the list (see crt_req_timeout_hdlr()).  Note that timeout_list is processed outside of lock on crt_ctx->cc_mutex.


Similarly with wait queue. Imagine some number of rpcs sitting in the wait queue, with max rpc in flight limit being hit already. If inflight drops by few (due to some inflight rpcs completing) then some rpcs can now be moved from wait queue and dispatched. To do that we then create a temp submit_list and move some rpcs from the wait_q onto that list.

once wait_q is processed and unlocked we then process the submit_list and dispatch each rpc.